streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -809,119 +782,42 @@ private void addRecordsToTasks(final ConsumerRecords<byte[], byte[]> records) {
         // until no task has any records left
         do {
-            totalProcessedEachRound = 0;
-            final Iterator<Map.Entry<TaskId, Task>> it = tasks.entrySet().iterator();
-            while (it.hasNext()) {
-                final Task task = it.next().getValue();
-                try {
-                    // we processed one record,
-                    // if more are buffered waiting for the next round
-
-                    // TODO: We should check for stream time punctuation right after each process call
-                    //       of the task instead of only calling it after all records being processed
-                    if (task.process()) {
-                        totalProcessedEachRound++;
-                        totalProcessedSinceLastMaybeCommit++;
-                    }
-                } catch (final ProducerFencedException e) {
-                    taskManager.closeZombieTask(task);
-                    it.remove();
-                }
+            processed = taskManager.process();
+            if (processed > 0) {
+                streamsMetrics.processTimeSensor.record(computeLatency() / (double) processed, timerStartedMs);
             }
+            totalProcessedSinceLastMaybeCommit += processed;
+
+            punctuate();
 
             if (recordsProcessedBeforeCommit != UNLIMITED_RECORDS &&
                 totalProcessedSinceLastMaybeCommit >= recordsProcessedBeforeCommit) {
                 totalProcessedSinceLastMaybeCommit = 0;
-                final long processLatency = computeLatency();
-                streamsMetrics.processTimeSensor.record(processLatency / (double) totalProcessedSinceLastMaybeCommit,
-                    timerStartedMs);
                 maybeCommit(timerStartedMs);
             }
-        } while (totalProcessedEachRound != 0);
-
-        // go over the tasks again to punctuate or commit
-        final RuntimeException e = taskManager.performOnActiveTasks(new TaskManager.TaskAction() {
-            private String name;
-            @Override
-            public String name() {
-                return name;
-            }
-
-            @Override
-            public void apply(final Task task) {
-                name = "punctuate";
-                maybePunctuateStreamTime(task);
-                if (task.commitNeeded()) {
-                    name = "commit";
-
-                    long beforeCommitMs = time.milliseconds();
-
-                    commitOne(task);
-
-                    if (log.isDebugEnabled()) {
-                        log.debug("{} Committed active task {} per user request in {}ms",
-                                logPrefix, task.id(), timerStartedMs - beforeCommitMs);
-                    }
-                }
-            }
-        });
-
-        if (e != null) {
-            throw e;
-        }
+        } while (processed != 0);
 
+        // go over the tasks again to maybe commit
```

Comment:
This call should be inside the do-while loop. More generally, today we have three commit calls and their purposes are a bit different:

1. Inside the do-while loop, after we made a single pass over each task trying to process one record from each, we should call `maybeCommitActiveTasks`, this is only for the case if user does trigger `context.commit` in the process function, in which case we should commit immediately before processing the next record.

2. Also inside the do-while loop, we may consider make a `maybeCommit` based on the total number of records processed, this is a heuristic work around to avoid time-based commit being called too late.

3. And outside the do-while loop, we call `maybeCommit` again, this is to avoid 2) above never called within the do-while loop. I feel that we can have a better way to tackle 2) / 3) together than what we did today, but that would be a different PR.